config/src/test/kotlin/org/springframework/security/config/web/servlet/OAuth2LoginDslTests.kt

```diff
@@ -113,6 +124,74 @@ class OAuth2LoginDslTests {
         }
     }
 
+    @Test
+    fun `oauth2Login when custom authentication details source then used`() {
+        this.spring
+            .register(CustomAuthenticationDetailsSourceConfig::class.java, ClientConfig::class.java)
+            .autowire()
+        mockkObject(CustomAuthenticationDetailsSourceConfig.AUTHENTICATION_DETAILS_SOURCE)
+        every {
+            CustomAuthenticationDetailsSourceConfig.AUTHENTICATION_DETAILS_SOURCE.buildDetails(any())
+        } returns Any()
+        mockkObject(CustomAuthenticationDetailsSourceConfig.AUTHORIZATION_REQUEST_REPOSITORY)
+        every {
+            CustomAuthenticationDetailsSourceConfig.AUTHORIZATION_REQUEST_REPOSITORY.removeAuthorizationRequest(any(), any())
+        } returns OAuth2AuthorizationRequest.authorizationCode()
+            .authorizationUri("/")
+            .clientId("clientId")
+            .redirectUri("/")
+            .attributes { attributes -> attributes[OAuth2ParameterNames.REGISTRATION_ID] = "google" }
+            .build()
+
+        this.mockMvc.post("/login/oauth2/code/google") {
+            param(OAuth2ParameterNames.CODE, "code")
+            param(OAuth2ParameterNames.STATE, "state")
+        }
+        .andExpect {
+            status { is3xxRedirection() }
+        }
+
+        verify(exactly = 1) { CustomAuthenticationDetailsSourceConfig.AUTHENTICATION_DETAILS_SOURCE.buildDetails(any()) }
+    }
+
+    @EnableWebSecurity
+    open class CustomAuthenticationDetailsSourceConfig : WebSecurityConfigurerAdapter() {
+
+        companion object {
+            val AUTHENTICATION_DETAILS_SOURCE: AuthenticationDetailsSource<HttpServletRequest, *> =
+                AuthenticationDetailsSource<HttpServletRequest, Any> { Any() }
+            val AUTHORIZATION_REQUEST_REPOSITORY = object: AuthorizationRequestRepository<OAuth2AuthorizationRequest> {
```

Comment:
I'd love a shorter alternative for this.

I tested with this one-liner: `val AUTHORIZATION_REQUEST_REPOSITORY = mockk<AuthorizationRequestRepository<in OAuth2AuthorizationRequest>>()`

But it behaved strangely on my machine. It passed when running the tests directly from Gradle, but, running from my IDE (IntelliJ) with this line would cause the spring `autowire()` call to hang indefinitely.

Plus, the `mockk` call felt redundant with the mocking in the actual test method.

Reply:
You can use a `HttpSessionOAuth2AuthorizationRequestRepository`. We do something similar in `OAuth2ClientDslTests`.